src/XrdClLazyDownload.cc

```diff
@@ -107,7 +107,16 @@ LDFile::Close(XrdCl::ResponseHandler *handler,
     AtomicBeg(m_mutex);
     AtomicCAS(m_is_open, true, false);
     AtomicEnd(m_mutex);
-    return m_fh.Close(handler, timeout);
+    if (m_fh.IsOpen()) {
```

Comment:
On Aug 24, 2015, at 3:10 PM, Brian Bockelman <notifications@github.com<mailto:notifications@github.com>> wrote:


In src/XrdClLazyDownload.cc<https://github.com/bbockelm/xrootd-lazy-download/pull/1#discussion_r37798306>:

> @@ -107,7 +107,16 @@ LDFile::Close(XrdCl::ResponseHandler *handler,
>      AtomicBeg(m_mutex);
>      AtomicCAS(m_is_open, true, false);
>      AtomicEnd(m_mutex);
> -    return m_fh.Close(handler, timeout);
> +    if (m_fh.IsOpen()) {


I don't think this is thread-safe. Shouldn't we be doing something based on m_is_open?

If m_fh.IsOpen() is thread-safe (which is appears to be) then this is too. Although it’s a little unpredictable whether you’ll get an error or not.

(I must admit that I think someone simultaneously calling close() from two different threads deserves bad things.)

Unfortunately AtomicCAS doesn’t provide a good way to determine whether the update to m_is_open happened or not. I think we’d have to switch to mutexes to actually use the value.

I'm worried about giving the correct error code in case the user does close() twice.